File: Data Structures & Algorithms/count-connected-components/submission-2.cpp
class Solution {
private:
    void dfs(int vertice, vector<vector<int>>& adjList, vector<int>& colors, int color) {
        if (colors[vertice]) return;

        colors[vertice] = color;

        for (auto adj : adjList[vertice])
            dfs(adj, adjList, colors, color);
    }

public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> adjList(n);

        for (auto e : edges) {
            adjList[e[0]].push_back(e[1]);
            adjList[e[1]].push_back(e[0]);
        }

        vector<int> colors(n);
        int color = 0;

        for (int vertice = 0; vertice < n; ++vertice)
            if (!colors[vertice])
                dfs(vertice, adjList, colors, ++color);

        return color;
    }
};

File: Data Structures & Algorithms/course-schedule-ii/submission-2.py
def dfs(ans, adj, colors, curr):
    if colors[curr] == 1:
        return True
    if colors[curr]:
        return False
    
    colors[curr] = 1

    for node in adj[curr]:
        if dfs(ans, adj, colors, node): 
            return True
    
    colors[curr] = 2
    ans.append(curr)
    

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(numCourses)]
        for p in prerequisites:
            adj[p[0]].append(p[1])

        colors = [0] * numCourses

        ans = []

        for i in range(numCourses):
            if dfs(ans, adj, colors, i):
                return []

        return ans